edge-runtime/lib/cjs.ts

```diff
@@ -0,0 +1,330 @@
+    }
+
+    if (subpath.length === 0 && pkgJson.main) {
+      return pkgJson.main
+    }
+
+    return subpath
+  }
+
+  memoizedPackageResolvers.set(packageJsonMatchedModule, resolveInPackage)
+
+  return resolveInPackage
+}
+
+function seedCJSModuleCacheAndReturnTarget(matchedModule: RegisteredModule, parent: Module) {
+  if (matchedModule.loaded) {
+    return matchedModule.filepath
+  }
+  const { source, filepath } = matchedModule
+
+  const mod = new Module(filepath)
+  mod.parent = parent
+  mod.filename = filepath
+  mod.path = dirname(filepath)
+  // @ts-expect-error - private untyped API
+  mod.paths = Module._nodeModulePaths(mod.path)
+  require.cache[filepath] = mod
+
+  try {
+    if (filepath.endsWith('.json')) {
+      Object.assign(mod.exports, parseJson(matchedModule))
+    } else {
+      const wrappedSource = `(function (exports, require, module, __filename, __dirname) { ${source}\n});`
+      const compiled = vm.runInThisContext(wrappedSource, {
+        filename: filepath,
+        lineOffset: 0,
+        displayErrors: true,
+      })
+      const modRequire = createRequire(pathToFileURL(filepath, { windows: false }))
+      compiled(mod.exports, modRequire, mod, filepath, dirname(filepath))
+    }
+    mod.loaded = matchedModule.loaded = true
+  } catch (error) {
+    throw new Error(`Failed to compile CJS module: ${filepath}`, { cause: error })
+  }
+
+  return filepath
+}
+
+// ideally require.extensions could be used, but it does NOT include '.cjs', so hardcoding instead
+const exts = ['.js', '.cjs', '.json']
+
+function tryWithExtensions(filename: string) {
+  let matchedModule = registeredModules.get(filename)
+  if (!matchedModule) {
+    for (const ext of exts) {
+      // require("./test") might resolve to ./test.js
+      const targetWithExt = filename + ext
+
+      matchedModule = registeredModules.get(targetWithExt)
+      if (matchedModule) {
+        break
+      }
+    }
+  }
+
+  return matchedModule
+}
+
+function tryMatchingWithIndex(target: string) {
+  let matchedModule = tryWithExtensions(target)
+  if (!matchedModule) {
+    // require("./test") might resolve to ./test/index.js
+    const indexTarget = join(target, 'index')
+    matchedModule = tryWithExtensions(indexTarget)
+  }
+
+  return matchedModule
+}
+
+export function registerCJSModules(baseUrl: URL, modules: Map<string, string>) {
+  const basePath = dirname(toPosixPath(fileURLToPath(baseUrl, { windows: false })))
+
+  for (const [filename, source] of modules.entries()) {
+    const target = join(basePath, filename)
+    registeredModules.set(target, { source, loaded: false, filepath: target })
+  }
+
+  if (!hookedIn) {
+    // @ts-expect-error - private untyped API
+    const original_resolveFilename = Module._resolveFilename.bind(Module)
+    // @ts-expect-error - private untyped API
+    Module._resolveFilename = (...args) => {
+      let target = args[0]
+      let isRelative = args?.[0].startsWith('.')
+
+      if (isRelative) {
```

Comment:
How do absolute paths work? Would an import have to start with `/virtual-root/...`?